File: Server/src/main/java/com/github/dr/rwserver/util/zip/zip/ZipDecoder.java
package com.github.dr.rwserver.util.zip.zip;

import com.github.dr.rwserver.game.GameMaps;
import com.github.dr.rwserver.struct.OrderedMap;
import com.github.dr.rwserver.struct.Seq;
import com.github.dr.rwserver.util.log.Log;
import com.github.dr.rwserver.util.zip.zip.realization.ZipEntry;
import com.github.dr.rwserver.util.zip.zip.realization.ZipFile;

import java.io.*;
import java.nio.charset.Charset;
import java.util.Enumeration;
import java.util.zip.GZIPInputStream;
import java.util.zip.ZipInputStream;


/**
 * 解码
 * @author Dr
 */
public class ZipDecoder {
    private ZipFile zipFile;
    private File file;

    public ByteArrayInputStream buffer;
    public DataInputStream stream;

    public ZipDecoder(File file) throws IOException {
        zipFile = new ZipFile(file);
        this.file = file;
    }

    public ZipDecoder(ZipFile zipFile) {
        this.zipFile = zipFile;
    }

    public ZipDecoder(byte[] bytes) {
        this.buffer = new ByteArrayInputStream(bytes);
    }

    public ZipDecoder(boolean bl,byte[] bytes) throws IOException {
        this.buffer = new ByteArrayInputStream(bytes);
        if (bl) {
            this.stream = new DataInputStream((InputStream)new BufferedInputStream((InputStream)new GZIPInputStream((InputStream)this.buffer)));
        } else {
            this.stream = new DataInputStream((InputStream)this.buffer);
        }
    }

    public OrderedMap<String, byte[]> getSpecifiedSuffixInThePackage(String endWith) {
        final OrderedMap<String,byte[]> data = new OrderedMap<>(8);
        try(ZipInputStream zis = new ZipInputStream(new FileInputStream(file), Charset.forName("GBK"))) {
            ZipEntry zipEntry;
            int len;
            InputStream in;
            final byte[] buffer = new byte[1024];
            try (ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream()) {
                for(Enumeration entries = zipFile.getEntries();entries.hasMoreElements();){
                    zipEntry = (ZipEntry)entries.nextElement();
                    in  =  zipFile.getInputStream(zipEntry);
                    while ((len = in.read(buffer)) != -1) {
                        byteArrayOutputStream.write(buffer, 0, len);
                    }
                    final String name = zipEntry.getName();
                    if (name.endsWith(endWith)) {
                        data.put(name.substring(0, name.length()-name.substring(name.lastIndexOf(".")).length()),byteArrayOutputStream.toByteArray());
                        //FileUtil.File(Data.Plugin_Cache_Path).toPath(name).writeFileByte(byteArrayOutputStream.toByteArray(),false);
                    }
                    byteArrayOutputStream.reset();
                }
            }
        } catch (IOException e) {
            Log.error(e);
        }
        return data;
    }

    public Seq<String> GetTheFileNameOfTheSpecifiedSuffixInTheZip(String endWith) {
        // Max 5M
        final int maxSize = 1024 * 1024 * 5;
        final Seq<String> data = new Seq<>(8);
        ZipEntry zipEntry;
        for(Enumeration entries = zipFile.getEntries();entries.hasMoreElements();){
            zipEntry = (ZipEntry)entries.nextElement();
            if (zipEntry.getSize() >= maxSize) {
                continue;
            }
            final String name = zipEntry.getName();
            if (name.endsWith(endWith)) {
                data.add(name.substring(0, name.length()-name.substring(name.lastIndexOf(".")).length()));
            }
        }
        return data;
    }

    public byte[] GetTheFileBytesOfTheSpecifiedSuffixInTheZip(final GameMaps.MapData mapData) {
        try(ZipInputStream zis = new ZipInputStream(new FileInputStream(file), Charset.forName("GBK"))) {
            ZipEntry zipEntry;
            int len;
            InputStream in;
            final byte[] buffer = new byte[1024];
            try (ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream()) {
                for(Enumeration entries = zipFile.getEntries();entries.hasMoreElements();){
                    zipEntry = (ZipEntry)entries.nextElement();
                    final String name = zipEntry.getName();
                    if (name.endsWith(mapData.getType()) && name.contains(mapData.mapFileName)) {
                        in  =  zipFile.getInputStream(zipEntry);
                        while ((len = in.read(buffer)) != -1) {
                            byteArrayOutputStream.write(buffer, 0, len);
                        }
                        return byteArrayOutputStream.toByteArray();
                    }
                }
            }
        } catch (IOException e) {
            Log.error(e);
        }
        return null;
    }

    public InputStreamReader getZipNameInputStream(String name) {
        try {
            @SuppressWarnings("unchecked")
            Enumeration<ZipEntry> entries = (Enumeration<ZipEntry>) zipFile.getEntries();
            ZipEntry ze;
            while (entries.hasMoreElements()) {
                ze = entries.nextElement();
                if (ze.isDirectory()) {
                    continue;
                } else {
                    if (ze.getName() != null) {
                        if (ze.getName().equals(name)) {
                            return new InputStreamReader(zipFile.getInputStream(ze),"UTF-8");
                        }
                    }
                }
            }
        } catch (Exception e) {
            Log.error(e);
        }
        return null;
    }

    public static InputStream getZipInputStream(InputStream in) throws Exception {
        return (InputStream)new BufferedInputStream((InputStream)new ZipInputStream(in));
    }
}

File: Server/src/main/kotlin/com/github/dr/rwserver/util/zip/zip/ZipDecoder.kt
/*
 * Copyright 2020-2022 RW-HPS Team and contributors.
 *
 * 此源代码的使用受 GNU AFFERO GENERAL PUBLIC LICENSE version 3 许可证的约束, 可以在以下链接找到该许可证.
 * Use of this source code is governed by the GNU AGPLv3 license that can be found through the following link.
 *
 * https://github.com/RW-HPS/RW-HPS/blob/master/LICENSE
 */

package com.github.dr.rwserver.util.zip.zip

import com.github.dr.rwserver.game.GameMaps.MapData
import com.github.dr.rwserver.struct.OrderedMap
import com.github.dr.rwserver.struct.Seq
import com.github.dr.rwserver.util.io.IoRead
import com.github.dr.rwserver.util.io.IoRead.MultiplexingReadStream
import com.github.dr.rwserver.util.log.Log.error
import com.github.dr.rwserver.util.log.exp.FileException
import org.apache.compress.ZipEntry
import org.apache.compress.ZipFile
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream
import java.nio.charset.Charset
import java.util.zip.ZipInputStream

/**
 * 解码
 * @author Dr
 */
class ZipDecoder {
    private var zipFile: ZipFile
    private lateinit var file: File

    constructor(file: File) {
        zipFile = ZipFile(file)
        this.file = file
    }

    constructor(zipFile: ZipFile) {
        this.zipFile = zipFile
    }

    /**
     * 获取ZIP内的指定后辍的文件名(无后辍)与bytes
     * @param endWith String
     * @return OrderedMap<String, ByteArray>
     */
    fun getSpecifiedSuffixInThePackage(endWith: String): OrderedMap<String, ByteArray> {
        val data = OrderedMap<String, ByteArray>(8)
        var zipEntry: ZipEntry
        try {
            MultiplexingReadStream().use { multiplexingReadStream ->
                val entries = zipFile.entries
                while (entries.hasMoreElements()) {
                    zipEntry = entries.nextElement() as ZipEntry
                    val nameCache = zipEntry.name
                    val name = nameCache.split("/").toTypedArray()[nameCache.split("/").toTypedArray().size - 1]
                    if (name.endsWith(endWith)) {
                        data.put(
                            name.substring(0, name.length - name.substring(name.lastIndexOf(".")).length),
                            multiplexingReadStream.readInputStreamBytes(zipFile.getInputStream(zipEntry))
                        )
                    }
                }
            }
        } catch (e: IOException) {
            error(e)
        }
        return data
    }

    /**
     * 获取ZIP内的指定后辍的文件名(全名)与bytes
     * @param endWith String
     * @return OrderedMap<String, ByteArray>
     */
    fun getSpecifiedSuffixInThePackageAllFileName(endWith: String): OrderedMap<String, ByteArray> {
        val data = OrderedMap<String, ByteArray>(8)
        try {
            ZipInputStream(FileInputStream(file), Charset.forName("GBK")).use {
                var zipEntry: ZipEntry
                MultiplexingReadStream().use { multiplexingReadStream ->
                    val entries = zipFile.entries
                    while (entries.hasMoreElements()) {
                        zipEntry = entries.nextElement() as ZipEntry
                        val nameCache = zipEntry.name
                        val name = nameCache.split("/").toTypedArray()[nameCache.split("/").toTypedArray().size - 1]
                        if (name.endsWith(endWith)) {
                            data.put(
                                name,
                                multiplexingReadStream.readInputStreamBytes(zipFile.getInputStream(zipEntry))
                            )
                        }
                    }
                }
            }
        } catch (e: IOException) {
            error(e)
        }
        return data
    }

    /**
     * 获取ZIP内满足后辍的文件名(无后辍)
     * @param endWith String
     * @return Seq<String>
     */
    fun getTheFileNameOfTheSpecifiedSuffixInTheZip(endWith: String): Seq<String> {
        // Max 5M
        val maxSize = 1024 * 1024 * 5
        val data = Seq<String>(8)
        var zipEntry: ZipEntry
        val entries = zipFile.entries
        while (entries.hasMoreElements()) {
            zipEntry = entries.nextElement() as ZipEntry
            if (zipEntry.size >= maxSize) {
                continue
            }
            val name = zipEntry.name
            if (name.endsWith(endWith)) {
                data.add(name.substring(0, name.length - name.substring(name.lastIndexOf(".")).length))
            }
        }
        return data
    }

    /**
     * 获取地图文件字节
     * @param mapData MapData
     * @return ByteArray
     */
    @Throws(Exception::class)
    fun getTheFileBytesOfTheSpecifiedSuffixInTheZip(mapData: MapData): ByteArray {
        ZipInputStream(FileInputStream(file), Charset.forName("GBK")).use {
            var zipEntry: ZipEntry
            val entries = zipFile.entries
            while (entries.hasMoreElements()) {
                zipEntry = entries.nextElement() as ZipEntry
                val name = zipEntry.name
                if (name.endsWith(mapData.type) && name.contains(mapData.mapFileName)) {
                    return IoRead.readInputStreamBytes(zipFile.getInputStream(zipEntry))
                }
            }
        }
        throw FileException("CANNOT_FIND_FILE")
    }

    fun getZipNameInputStream(name: String): InputStream? {
        try {
            val entries = zipFile.entries
            var ze: ZipEntry
            while (entries.hasMoreElements()) {
                ze = entries.nextElement()
                if (!ze.isDirectory) {
                    if (ze.name == name) {
                        return zipFile.getInputStream(ze)
                    }
                }
            }
        } catch (e: Exception) {
            error(e)
        }
        return null
    }
}